### PR TITLE
Avoid the media stream was disposed before it's finished.

### LINKED ISSUE #17859: The stream was disposed before its real end
### Describe the bug

creatingStream was disposed , which made the event hander could not manipulate the stream.

### Orchard Core version
latest main

### To Reproduce

1. Go to CreateFileFromStreamAsync of DefaultMediaFileStore
2. See error 
The stream was not opened for writing.
or disposed

### Logs and screenshots

----------------
```
2025-04-27 16:06:36.0753|Default|00-2f2c2e1e33d2f3a42eac04632ee95bfd-38307f23cf03942c-00||OrchardCore.Media.Controllers.AdminController|ERROR|An error occurred while uploading a media OrchardCore.FileStorage.FileStoreException: Cannot create file 'zg2/Andy-UX.docx'.
 ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.String'.
   at Microsoft.AspNetCore.Http.ReferenceReadStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Stream.<CopyToAsync>g__Core|27_0(Stream source, Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
   at OrchardCore.FileStorage.FileSystem.FileSystemStore.CreateFileFromStreamAsync(String path, Stream inputStream, Boolean overwrite)
   --- End of inner exception stack trace ---
   at OrchardCore.FileStorage.FileSystem.FileSystemStore.CreateFileFromStreamAsync(String path, Stream inputStream, Boolean overwrite)
   at OrchardCore.Media.Core.DefaultMediaFileStore.CreateFileFromStreamAsync(String path, Stream inputStream, Boolean overwrite)
   at OrchardCore.Media.Controllers.AdminController.<>c__DisplayClass19_0.<<Upload>b__1>d.MoveNext()    at OrchardCore.FileStorage.FileSystem.FileSystemStore.CreateFileFromStreamAsync(String path, Stream inputStream, Boolean overwrite)
   at OrchardCore.Media.Core.DefaultMediaFileStore.CreateFileFromStreamAsync(String path, Stream inputStream, Boolean overwrite)
   at OrchardCore.Media.Controllers.AdminController.<>c__DisplayClass19_0.<<Upload>b__1>d.MoveNext()
```

### PR DESCRIPTION
**Pull request by infofromca**
Fix #17859

### CODE REVIEW

**Comment by gvkries** [comment]
Maybe we should reconsider this more. Disposing the `inputStream` is unexpected for the callers of `CreateFileFromStreamAsync` and we probably should return the stream created by the event handlers.

**Comment by infofromca** [comment]
I found this:
```
 mediaFilePath = await _mediaFileStore.CreateFileFromStreamAsync(mediaFilePath, stream);

 var mediaFile = await _mediaFileStore.GetFileInfoAsync(mediaFilePath);

 // The .NET AWS SDK, and only that from the built-in ones (but others maybe too), disposes
 // the stream. There's no better way to check for that than handling the exception. An
 // alternative would be to re-read the file for every other storage provider as well but
 // that would be wasteful.
 try
 {
     stream.Position = 0;
 }
 catch (ObjectDisposedException)
 {
     stream = null;
 }

```
so, it seems we do not need  inputStream = null;

File: src/OrchardCore/OrchardCore.Media.Core/DefaultMediaFileStore.cs [hunk 1]
```diff
@@ -169,24 +169,34 @@ public virtual async Task<string> CreateFileFromStreamAsync(string path, Stream
...
 
                 return await _fileStore.CreateFileFromStreamAsync(context.Path, outputStream, overwrite);
             }
             finally
             {
-                // This disposes the last outputStream.
-                outputStream?.Dispose();
+                if (inputStream != outputStream)
+                {
+                    inputStream = outputStream;
```
Comment (gvkries) [inline src/OrchardCore/OrchardCore.Media.Core/DefaultMediaFileStore.cs:198]: You can't 'return' the stream this way. If you want to update the reference for the caller as well, you must pass the argument by reference, e.g., `CreateFileFromStreamAsync(string path, ref Stream inputStream, bool overwrite = false)`. However, this requires changes to the interface contract.

**Review by gvkries — CHANGES_REQUESTED** [review]

**Comment by sebastienros** [comment]
I don't understand why we need to return a stream. We return a file path, so the user could just open the stream with this information if needed.

**Comment by infofromca** [comment]
@gvkries and @sebastienros I made  a new PR at #17878. 
Pick the one you like

**Comment by gvkries** [comment]
> I don't understand why we need to return a stream. We return a file path, so the user could just open the stream with this information if needed.

The issue is that event handlers are allowed to return a new stream, which may alter the content of the original input. Since the callers of the method are not aware of this behavior and may reuse the `inputStream` afterwards (e.g., [in this example](https://github.com/OrchardCMS/OrchardCore/blob/97884630a19b14f929a1181368f6df9a9855f4d8/src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs#L258)), it gets disposed to prevent the caller from using outdated contents. However, disposing is also not the responsibility of this method. Therefore, returning an updated stream makes sense to me.

**Comment by infofromca** [comment]
> > I don't understand why we need to return a stream. We return a file path, so the user could just open the stream with this information if needed.
> 
> The issue is that event handlers are allowed to return a new stream, which may alter the content of the original input. Since the callers of the method are not aware of this behavior and may reuse the `inputStream` afterwards (e.g., [in this example](https://github.com/OrchardCMS/OrchardCore/blob/97884630a19b14f929a1181368f6df9a9855f4d8/src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs#L258)), it gets disposed to prevent the caller from using outdated contents. However, disposing is also not the responsibility of this method. Therefore, returning an updated stream makes sense to me.

I agree. otherwise await PreCacheRemoteMedia(mediaFile, stream); will cache different media.

File: src/OrchardCore.Modules/OrchardCore.Lists/RemotePublishing/MetaWeblogHandler.cs [hunk 2]
```diff
@@ -153,15 +155,30 @@ private async Task<XRpcStruct> MetaWeblogNewMediaObjectAsync(string userName, st
 
         var directoryName = Path.GetDirectoryName(name);
         var filePath = _mediaFileStore.Combine(directoryName, Path.GetFileName(name));
-        Stream stream = null;
+        MemoryStream stream = null;
+        (string outputPath, Stream outputStream)? output = null;
```
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.Lists/RemotePublishing/MetaWeblogHandler.cs:159]: You don't need to make this nullable, just test the inner fields for null.

File: src/OrchardCore.Modules/OrchardCore.Lists/RemotePublishing/MetaWeblogHandler.cs [hunk 3]
```diff
@@ -153,15 +155,31 @@ private async Task<XRpcStruct> MetaWeblogNewMediaObjectAsync(string userName, st
 
         var directoryName = Path.GetDirectoryName(name);
         var filePath = _mediaFileStore.Combine(directoryName, Path.GetFileName(name));
-        Stream stream = null;
+        MemoryStream stream = null;
+        (string outputPath, Stream outputStream) output = (null, null); // Initialize the tuple to avoid CS0170
```
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.Lists/RemotePublishing/MetaWeblogHandler.cs:159]: I would prefer to use `default` instead:

```suggestion
        (string outputPath, Stream outputStream) output = default;
```

File: src/OrchardCore/OrchardCore.Media.Abstractions/IMediaFileStore.cs [hunk 4]
```diff
@@ -13,4 +14,28 @@ public interface IMediaFileStore : IFileStore
...
+    /// <item>
+    /// <description><c>outputStream</c>: The resulting stream, which may be the original input stream or a modified stream if any event handlers altered it.</description>
+    /// </item>
+    /// </list>
+    /// </returns>
+    /// <remarks>
+    /// If the specified path contains one or more directories, then those directories are
+    /// created if they do not already exist.
+    /// </remarks>
+    Task<(string outputPath, Stream outputStream)> CreateMediaFileFromStreamAsync(MediaCreatingContext context, Stream inputStream, bool overwrite = false) => Task.FromResult((context.Path, inputStream));
```
Comment (gvkries) [inline src/OrchardCore/OrchardCore.Media.Abstractions/IMediaFileStore.cs:39]: To change the interface in a backward compatible way, the default implementation must actually work. So you need to call the existing method here. E.g.

```csharp
async Task<(string outputPath, Stream outputStream)> CreateMediaFileFromStreamAsync(MediaCreatingContext context, Stream inputStream, bool overwrite = false)
{
    var outputPath = await CreateFileFromStreamAsync(context.Path, inputStream, overwrite);
    return (outputPath, inputStream);
}
```

**Comment by sebastienros** [comment]
This is too complicated, and there is no current implementation of the special event handler. I would suggest to delete this media handler event interface and complete the existing one with 

```cs
    Task MediaCreatingFileAsync(MediaCreatingContext context);
    Task MediaCreatedFileAsync(MediaCreatedContext context);
```

This way the event handlers can just replace or create new file without is needing to manage the intermediate streams.

@infofromca would it still let you create/update the media when they are uploaded (if this is your scenario)?

**Comment by infofromca** [comment]
I would like to use special event handler for many reasons, for example, validation and sanitation

**Comment by gvkries** [comment]
I advocated for this solution in yesterday’s meeting, but the general consensus leaned toward removing the specialized `IMediaCreatingEventHandler` and opting for a simpler approach.

This alternative would allow event handlers to work directly with the file after it has been uploaded to the store. While this simplifies the implementation, I revisited the original PR (#5098) and still prefer a solution that enables modifying the incoming stream *before* it’s written to the store.

The simpler approach has some trade-offs:

- It introduces additional I/O operations,
- It may result in slower performance,
- And it could lead to higher costs (e.g., on AWS or Azure).

I remain convinced that returning the stream is the more efficient and flexible solution in this case. However, given the current direction, I would suggest proceeding with the simpler implementation proposed in #17878.

**Comment by gvkries** [comment]
We'll take #17878 instead of this.
